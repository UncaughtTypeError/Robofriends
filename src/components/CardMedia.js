import React from 'react';
import ErrorBoundryCardMedia from '../components/ErrorBoundryCardMedia';

import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import CardMedia from '@material-ui/core/CardMedia';
import CircularProgress from '@material-ui/core/CircularProgress';
import Tooltip from '@material-ui/core/Tooltip';

const styles = {
    tooltip: {
        fontSize: 15,
    },
    media: {
        // ⚠️ object-fit is not supported by IE 11.
        objectFit: 'cover',
        minWidth: 300,
        minHeight: 350,
        zIndex: 2,
        position: 'relative',
    },
    mediaTemp: {
        opacity: '.2',
        filter: 'grayscale(1)',
    },
    mediaLoading: {
        display: 'none',
    },
    progress: {
        position: 'absolute',
        left: 0,
        right: 0,
        margin: 'auto',
        top: 0,
        bottom: 0,
        zIndex: 1,    
        color: '#6f6f6f',
    },
};

class CardMediaWrapper extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            imgUrl: null,
            tempUrl: `https://robohash.org/${this.props.id}?300x350`,
            elID: 'CardMediaWrapper'+this.props.id,
            imgLoaded: false,
            imgTimeout: null,
        }
    }

    async componentDidMount() {  

        let imgID = Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5);

        try {
            const response = await fetch(`https://robohash.org/${imgID}?300x350`);
            if (!response.ok) {
                console.error('(http error)',response.statusText);
            } 
            const url = await response.url; 

            this.setState({ imgUrl: url });  

            let elID = document.getElementById(this.state.elID);
            elID.onload = () => {
                this.setState({ imgLoaded: true });  
                //console.log(this.state);
            }
            elID.onerror = (error) => {
                this.setState({ imgUrl: this.state.tempUrl }); 
                console.error(error);
                //throw new Error(error);
            }

        } catch (error) {
            console.error('(network error)',error);
        };

        if(!this.state.imgLoaded) {
            this.setState({imgTimeout: setTimeout(() => { this.forceSetImg() }, 30000)});
        }
    }

    forceSetImg() {
        if(!this.state.imgLoaded) { // img still hasn't loaded after timeout, force state update
            this.setState({ 
                imgLoaded: true,
                imgUrl: this.state.tempUrl,
            });  
            console.error('(timed out)',this.state);
        }
    }

    render() {
        const { classes, name } = this.props;

        return(
            <ErrorBoundryCardMedia name={name} id={this.state.elID}>
                <Tooltip classes={{ tooltip: classes.tooltip }} title={this.state.imgUrl ? name : 'Loading...'} placement="top">
                    <CardMedia
                    component="img"
                    alt={name}
                    className={this.state.imgLoaded ? classes.media : `${classes.media} ${classes.mediaTemp}`}
                    height="350"
                    image={this.state.imgUrl ? this.state.imgUrl : this.state.tempUrl}
                    aria-label={name}
                    id={this.state.elID}
                    />
                </Tooltip>
                { !this.state.imgLoaded ? <CircularProgress className={classes.progress} /> : null}
            </ErrorBoundryCardMedia>
        );
    }

}

CardMediaWrapper.propTypes = {
    classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(CardMediaWrapper);