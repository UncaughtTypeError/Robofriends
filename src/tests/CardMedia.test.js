import React from 'react';
import { shallow, mount, render } from 'enzyme';

import CardMedia from '../components/CardMedia';

it('expect to render full CardMedia component tree', () => {

    //expect(shallow(<CardGrid />).length).toEqual(1);

    const mockProps = {
        mockName: 'robo',
        mockID: 1
    },
    { mockName, mockID } = mockProps;

    expect(render(<CardMedia name={mockName} id={mockID} />)).toMatchSnapshot();
});