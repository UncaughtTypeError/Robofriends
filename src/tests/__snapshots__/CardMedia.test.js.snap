// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render full CardMedia component tree 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "alt": "robo",
      "aria-label": "robo",
      "class": "MuiCardMedia-root-15 MuiCardMedia-media-16 CardMediaWrapper-media-2 CardMediaWrapper-mediaTemp-3",
      "height": "350",
      "id": "CardMediaWrapper1",
      "src": "https://robohash.org/1?size=300x350",
      "title": "Loading...",
    },
    "children": Array [],
    "name": "img",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": Object {
      "attribs": Object {
        "class": "MuiCircularProgress-root-17 MuiCircularProgress-colorPrimary-20 MuiCircularProgress-indeterminate-19 CardMediaWrapper-progress-5",
        "role": "progressbar",
        "style": "width:40px;height:40px",
      },
      "children": Array [
        Object {
          "attribs": Object {
            "class": "MuiCircularProgress-svg-22",
            "viewBox": "22 22 44 44",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "MuiCircularProgress-circle-23 MuiCircularProgress-circleIndeterminate-25",
                "cx": "44",
                "cy": "44",
                "fill": "none",
                "r": "20.2",
                "stroke-width": "3.6",
              },
              "children": Array [],
              "name": "circle",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "cx": undefined,
                "cy": undefined,
                "fill": undefined,
                "r": undefined,
                "stroke-width": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "cx": undefined,
                "cy": undefined,
                "fill": undefined,
                "r": undefined,
                "stroke-width": undefined,
              },
            },
          ],
          "name": "svg",
          "namespace": "http://www.w3.org/2000/svg",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "viewBox": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "viewBox": undefined,
          },
        },
      ],
      "name": "div",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": [Circular],
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {
        "class": undefined,
        "role": undefined,
        "style": undefined,
      },
      "x-attribsPrefix": Object {
        "class": undefined,
        "role": undefined,
        "style": undefined,
      },
    },
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
        Object {
          "attribs": Object {
            "class": "MuiCircularProgress-root-17 MuiCircularProgress-colorPrimary-20 MuiCircularProgress-indeterminate-19 CardMediaWrapper-progress-5",
            "role": "progressbar",
            "style": "width:40px;height:40px",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "MuiCircularProgress-svg-22",
                "viewBox": "22 22 44 44",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "MuiCircularProgress-circle-23 MuiCircularProgress-circleIndeterminate-25",
                    "cx": "44",
                    "cy": "44",
                    "fill": "none",
                    "r": "20.2",
                    "stroke-width": "3.6",
                  },
                  "children": Array [],
                  "name": "circle",
                  "namespace": "http://www.w3.org/2000/svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "cx": undefined,
                    "cy": undefined,
                    "fill": undefined,
                    "r": undefined,
                    "stroke-width": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "cx": undefined,
                    "cy": undefined,
                    "fill": undefined,
                    "r": undefined,
                    "stroke-width": undefined,
                  },
                },
              ],
              "name": "svg",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "viewBox": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "viewBox": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": null,
          "prev": [Circular],
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "role": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "role": undefined,
            "style": undefined,
          },
        },
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "alt": undefined,
      "aria-label": undefined,
      "class": undefined,
      "height": undefined,
      "id": undefined,
      "src": undefined,
      "title": undefined,
    },
    "x-attribsPrefix": Object {
      "alt": undefined,
      "aria-label": undefined,
      "class": undefined,
      "height": undefined,
      "id": undefined,
      "src": undefined,
      "title": undefined,
    },
  },
  "1": Object {
    "attribs": Object {
      "class": "MuiCircularProgress-root-17 MuiCircularProgress-colorPrimary-20 MuiCircularProgress-indeterminate-19 CardMediaWrapper-progress-5",
      "role": "progressbar",
      "style": "width:40px;height:40px",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "MuiCircularProgress-svg-22",
          "viewBox": "22 22 44 44",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "MuiCircularProgress-circle-23 MuiCircularProgress-circleIndeterminate-25",
              "cx": "44",
              "cy": "44",
              "fill": "none",
              "r": "20.2",
              "stroke-width": "3.6",
            },
            "children": Array [],
            "name": "circle",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "cx": undefined,
              "cy": undefined,
              "fill": undefined,
              "r": undefined,
              "stroke-width": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "cx": undefined,
              "cy": undefined,
              "fill": undefined,
              "r": undefined,
              "stroke-width": undefined,
            },
          },
        ],
        "name": "svg",
        "namespace": "http://www.w3.org/2000/svg",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "viewBox": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "viewBox": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": Object {
      "attribs": Object {
        "alt": "robo",
        "aria-label": "robo",
        "class": "MuiCardMedia-root-15 MuiCardMedia-media-16 CardMediaWrapper-media-2 CardMediaWrapper-mediaTemp-3",
        "height": "350",
        "id": "CardMediaWrapper1",
        "src": "https://robohash.org/1?size=300x350",
        "title": "Loading...",
      },
      "children": Array [],
      "name": "img",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": [Circular],
      "parent": null,
      "prev": null,
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {
        "alt": undefined,
        "aria-label": undefined,
        "class": undefined,
        "height": undefined,
        "id": undefined,
        "src": undefined,
        "title": undefined,
      },
      "x-attribsPrefix": Object {
        "alt": undefined,
        "aria-label": undefined,
        "class": undefined,
        "height": undefined,
        "id": undefined,
        "src": undefined,
        "title": undefined,
      },
    },
    "root": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "attribs": Object {
            "alt": "robo",
            "aria-label": "robo",
            "class": "MuiCardMedia-root-15 MuiCardMedia-media-16 CardMediaWrapper-media-2 CardMediaWrapper-mediaTemp-3",
            "height": "350",
            "id": "CardMediaWrapper1",
            "src": "https://robohash.org/1?size=300x350",
            "title": "Loading...",
          },
          "children": Array [],
          "name": "img",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": null,
          "prev": null,
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "alt": undefined,
            "aria-label": undefined,
            "class": undefined,
            "height": undefined,
            "id": undefined,
            "src": undefined,
            "title": undefined,
          },
          "x-attribsPrefix": Object {
            "alt": undefined,
            "aria-label": undefined,
            "class": undefined,
            "height": undefined,
            "id": undefined,
            "src": undefined,
            "title": undefined,
          },
        },
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "role": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "role": undefined,
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 2,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
